File: plugins/devkit/devkit-java-tests/testData/threadingModelHelper/SubtypingPolymorphism.java
import com.intellij.util.concurrency.annotations.RequiresReadLock;
import com.intellij.util.concurrency.ThreadingAssertions;

public class SubtypingPolymorphism {
  public void testMethod() {
    Service[] services = {new FileService(), new UIService(), new DatabaseService()};
    for (Service service : services) {
      service.execute();
    }
  }
}

interface Service {
  void execute();
}

class FileService implements Service {
  @Override
  @RequiresReadLock
  public void execute() { }
}

class UIService implements Service {
  @Override
  public void execute() {
    ThreadingAssertions.assertEventDispatchThread();
  }
}

class DBService implements Service {
  @Override
  public void execute() {
    ThreadingAssertions.assertReadAccess();
  }
}

File: plugins/devkit/devkit-kotlin-tests/testData/threadingModelHelper/SubtypingPolymorphism.kt
import com.intellij.util.concurrency.annotations.RequiresReadLock
import com.intellij.util.concurrency.annotations.RequiresEdt
import com.intellij.util.concurrency.ThreadingAssertions

class SubtypingPolymorphism {
  fun testMethod() {
    val services: List<BaseService> = listOf(FileService(), UIService(), DBService())
    services.forEach { it.execute() }
  }
}

interface BaseService {
  fun execute()
}

class FileService : BaseService {
  @RequiresReadLock
  override fun execute() {
  }
}

class UIService : BaseService {
  override fun execute() {
    ThreadingAssertions.assertEventDispatchThread()
  }
}

class DBService : BaseService {
  override fun execute() {
    ThreadingAssertions.assertReadAccess()
  }
}
